### ISSUE TITLE
Integration of pdf annotation extraction in Zotero (from zotfile)

### ISSUE THREAD

**Issue by jlegewie** [zotero/zotero#1018]
Hi,

I just wanted to open an issue for a discussion about this. Presumably this is not for 5.0 but if the Zotero devs are interested in this, maybe 5.1.

The idea is to integrate the annotation extraction feature from zotfile in zotero. I think it's such an essential feature for reference managers and would be awesome for zotero. It would, however, require some work from the zotero devs because I won't be able to do everything in a pull request. The feature in Zotfile is pretty mature (e.g. unit tests for extraction quality exist but not with continuous integration). Still some parts might not be up to Zotero standards and it would require data base and UI changes. But the biggest burden for zotero core might be future maintenance and user support. You would ship zotero with a modified version of pdf.js, which is a huge library with rapid development. Updating to the most recent version can be a pain (mainly because I am using a modified version of pdf.js for this feature). Just saying that there are drawbacks.
#### Proposed UI

![zotero-annotations](https://cloud.githubusercontent.com/assets/1292401/15653277/0c9b9a0a-265a-11e6-9fad-b23c8e7857d2.jpg)

Currently, the attachment info pane has no tabs and just shows the left tab (without the tabs bar, of course). Instead, when the user selects an attachment, the pane could show three tabs with "Info" (same as before), Annotations with a list of extracted annotations, and Outline with the extracted outline.
The Annotation tab would be updated in the background so it would always shows the current annotations in the file. The links allow the user to open the pdf attachment on the page with the highlighted text. The "Create separate note with annotations" button would create a child note with the extracted annotations. This is important because the annotation tab itself can not be edited. Creating a note would allow the user to work with and edit the annotations in a note by adding comments etc. Zotfile only creates such a note and doesn't show the always updated list of annotations in a separate non-editable tab. Finally, the outline tab shows the pdf outline. Clicking on a section in the outline would open the pdf on that page. Of course, this only works for pdfs with an outline.

**Comment by dstillman** [comment]
Thanks for starting this discussion. I've thought for a while that it could make sense to migrate some ZotFile features into Zotero core.

So, yeah, my assumption has always been that, if we did support extracted annotations, we'd want them to be virtual, which I think is basically what you're describing: they'd reflect what was currently in the PDF, not be a manual operation. (Not sure what pdf.js allows, but seems like it might even be possible for them to be writable from the note view, no?)

I'd imagine a different view for the Annotations tab, closer to the existing Notes tab. We've also discussed showing annotations as (virtual) notes under the item in the middle pane.

Re: outlines, that doesn't seem to me like something that belongs in Zotero. I think that makes more sense just being in the PDF viewer.

**Comment by jlegewie** [comment]
Just a couple of thoughts:

- Agree about the virtual aspect but also think that an option to create note from annotations is very important because it makes it possible to work with the extracted text (add own thoughts etc).
- Not 100% sure what you mean with "writable" but I don't think that is possible.
- A benefit of having the outline is that you can directly jump to specific sections in the pdf. Mendeley has a feature like that. But the downside is that most pdf don't include an outline. Depends on the publisher and many just don't do it.

**Comment by dstillman** [comment]
By writable I mean the ability to edit your own annotations (as opposed to your highlights) from the virtual entry. But I realize now that your screenshot actually shows a highlight, not an annotation — do you support extraction of both?

The outline part still doesn't make sense to me. The PDF viewer should let you do this, and even pdf.js supports it. Not sure why we would duplicate that functionality — a tab in Zotero is no different from a sidebar in the PDF viewer.

**Comment by jlegewie** [comment]
Yes, the screenshot shows highlighted text and not a text annotation (I think they are called "Text" and "Highlight" annotations in pdfs). Zotfile supports both (and underlined text as well). Text annotations are pretty easy. It's possible to extract them with the pdf.js API without rendering the actual pdf (so it's also very fast). Extracting highlighted text is much more complicated and not possible with the pdf.js API. Zotfile uses a modified version of pdf.js to extract highlighted and underlined text. It's also more computationally intense because zotfile has to render all pages with annotations in a hidden browser to check which texts falls into the highlighted area (defined by annotation quadpoints in pdf terminology).

re writable: I don't think pdf.js supports any modifications to existing annotations. The official statement is that "PDF.js won't support editing features, it's only a reader" (e.g. timvandermeij comment [here](https://github.com/mozilla/pdf.js/issues/3767)). Would be pretty nice for text annotations though. Not sure how hard it would be to implement this.

**Comment by gracile-fr** [comment]
<blockquote>The outline part still doesn't make sense to me. The PDF viewer should let you do this, and even pdf.js supports it. Not sure why we would duplicate that functionality — a tab in Zotero is no different from a sidebar in the PDF viewer.</blockquote>It's quite handy to have the outline of a document in Zotero, especially because it's an active outline (i.e. with direct links as Joscha said). Also, the way it works with Zotfile at the moment, i.e. the fact that a note is created, makes possible to open it in another window and resize it as needed.

**Comment by dstillman** [comment]
@gracile-fr: This isn't about annotations — this is about the outline from the PDF, which (when present in a PDF, which is rare) you can view in a sidebar in pretty much any PDF viewer, including pdf.js in Firefox. I'm suggesting that it doesn't make sense to duplicate that functionality.

**Comment by dstillman** [comment]
The main difference, I suppose, is the ability to browse through PDFs in your library, leaving the outline view selected, and see the outline of each, without having to open each one. Not sure how much that matters for real-world usage.

**Comment by sojusnik** [comment]
What is the current status on this essential feature? Any progress recently?

**Comment by dstillman** [comment]
@sojusnik: No ETA, but we're working on this.

**Comment by StoltHD** [comment]
IT would be really nice if you could prioritize this, because the pdf.js version used in Zotfile is 5 years old and do not support the latest versions of the pdf standards...

And I find that the newer version 1.7 is used more often now than the "older" versions...
It would be nice if we could keep on extracting annotations, selections and comments from newer pdf's in Zotero instead of searching for new software...

And maybe you should cooperate a little with the Tropy team, they need to implement the same functionality in that software...

**Comment by jlegewie** [comment]
Hi @dstillman, I just wanted to check in on the status of this. Is this on the horizon in 2020? If not, I would consider options for getting annotation extraction in zotfile updated but I can't promise anything on that. Just asking before I even consider.

**Comment by dstillman** [comment]
Yes, we should have something on this in the next few months.

**Comment by derWalter** [comment]
Hey guys,

may I ask what the status on this topic is?
I guess if you make single annotations/highlighted or underlined text
addable per quick search in the word processors, we could even do
a KickStarter campaign for this, as people asking for this for nearly fifteen years now 😄 
catchphrase: "knowledge organization"

@dstillman several years ago you asked for reasons why to duplicate functionality

tl;dr : having the functionality close by where you need it, without having to change setting

the most easy answer for me personally is: I do not need to switch between programs,
which makes workflow more smooth and easy.

I guess thats the reason why all toolshops own the same kind of tools,
so they dont have to borrow them all the time from each other and
therefore can work smoother and faster 😃 

 If I could only read, underline/highlight, annotate
pdf's directly in zotero and could use these individually in my word processor, I would not
have to deal with and pay for an other piece of software (eg. Adobe Acrobat).

**Comment by derWalter** [comment]
Hey guys,

I just wanted to check back on this.
Sadly not even a comment from official side :/



Right now, I am looking for an ebookreader, with highlighting capability and
the option to automatically sync my changes to Zotero.
(I know about sync+Zotfile 👍)

It would be awesome to just buy ANY tablet, install Zotero and start working (highlighting/annotating),
go back to my Desktop and continue where I just left off.

Right now I need to research the right apps for the task,
learn how to setup everything, get it to work, maintain three different programs
on different platforms. Yes its working, but its time consuming.

If it would be all settled in Zotero (and have it finally on android) I would
just buy a Linux tablet and kick it off :)


But nevertheless I am so glad we have Zotero out there and
are able to use it for free! Thank you so much guys for your great work and effort,
I hope that Zotero keeps going, like its been since its beginning!


🌹 🌹 🌹 🌹 🌹 🌹 🌹 🌹

**Comment by Victor239** [comment]
It seems to be available [in dev builds now](https://www.zotero.org/support/pdf_reader_preview).

**Comment by dstillman** [comment]
Yes, this is no longer needed due to the [new PDF functionality](https://www.zotero.org/support/pdf_reader_preview) available in the Zotero beta, which allows you to extract both embedded file annotations and Zotero annotations directly into Zotero notes.

**Comment by derWalter** [comment]
😮  wow ... an open source project, that actually listens to user wishes, how cool is that????

thanks guys for letting me/us know about it!


if they also manage to separate the annotations into individual notes, or a new kind of "annotation" item, which can be used in a word processor like a citation (search for it and quote/paste it including the right citation), then the program would be finished, at least from my perspective 😄

**Comment by dstillman** [comment]
@derWalter: You can insert Zotero notes containing Zotero annotations directly into your word processor documents, with active Zotero citations. See the linked announcement.
